arxiv_paper_tool.py

HarikrishnanK9 opened the pull request:
Custom tool for fetching data from Arxiv.Will be helpful for Crewai community to utilize this data for their projects and experiments.This Integration is crucial for learners and enthusiasts as well as researchers to identify and explore advanced research outcomes from Arxiv papers on various topics with the help of Crewai

joaomdmoura commented:
Disclaimer: This review was made by a crew of AI Agents.

# Code Review Comment for `arxiv_paper_tool.py`

## Overview
The implementation of the `ArxivPaperTool` for fetching and downloading ArXiv papers is commendable, showcasing a well-structured approach to utilizing the ArXiv API. However, several key improvements can enhance code quality, maintainability, and overall functionality.

## Positive Aspects
1. **Type Annotations and Data Validation**: The integration of type hints and Pydantic models aids in clarity and error prevention.
2. **Effective Logging and Error Handling**: The focus on logging enhances the tool's debuggability during both development and production use.
3. **Politeness to API**: The inclusion of rate limiting via sleep intervals exhibits a responsible approach to API usage.

## Identified Issues and Recommendations

### 1. **Duplicate Imports**
The `os` module is imported twice. This should be corrected to improve code cleanliness:
```python
import os  # Remove duplicate occurrences
```

### 2. **Unused Imports**
Eliminate the imports from `crewai` since they are not utilized in this context:
```python
# Remove these lines
from crewai import Agent, Task, Crew, LLM
from crewai.process import Process
```

### 3. **Use of Constants**
Hardcoded values should be defined as constants within the class for better maintainability:
```python
class ArxivPaperTool(BaseTool):
    BASE_API_URL = "http://export.arxiv.org/api/query"
    SLEEP_DURATION = 1
```
This allows for easier modifications and encourages clarity regarding the purpose of these values.

### 4. **Improvement in XML Namespace Handling**
Reduce redundancy with XML namespace handling by employing a single method:
```python
@staticmethod
def _get_element_text(entry, element_name):
    return entry.find(f'{self.ATOM_NAMESPACE}{element_name}').text.strip() if entry.find(f'{self.ATOM_NAMESPACE}{element_name}') is not None else None
```

### 5. **Refined PDF URL Extraction**
Optimize the function for extracting PDF links:
```python
def _extract_pdf_url(self, entry) -> Optional[str]:
    return next((link.attrib['href'] for link in entry.findall(f'{self.ATOM_NAMESPACE}link') if 'pdf' in link.attrib.get('title', '').lower()), None)
```
This cleanly uses the next generator pattern to find the desired link.

### 6. **Separating Formatting Logic**
Create a dedicated method for generating formatted output from paper metadata:
```python
def _format_paper_result(self, paper: dict) -> str:
    summary = (paper['summary'][:self.SUMMARY_TRUNCATE_LENGTH] + '...') if len(paper['summary']) > self.SUMMARY_TRUNCATE_LENGTH else paper['summary']
    return f"Title: {paper['title']}\nAuthors: {', '.join(paper['authors'])}\nPublished: {paper['published_date']}\nPDF: {paper['pdf_url'] or 'N/A'}\nSummary: {summary}"
```

### 7. **Input Validation**
Enhance validation on `max_results` and other inputs:
```python
class ArxivToolInput(BaseModel):
    max_results: int = Field(5, ge=1, le=100, description="Max results to fetch; must be between 1 and 100")
```

### 8. **File Safety Enhancements**
Ensure the validity of file paths to mitigate security risks:
```python
def _validate_save_path(self, save_dir: str) -> None:
    if not os.path.isabs(save_dir):
        save_dir = os.path.abspath(save_dir)
    if not os.path.exists(save_dir):
        os.makedirs(save_dir, exist_ok=True)
```

### 9. **Documentation Improvements**
Enhance the code with detailed docstrings for classes and methods, improving user comprehension and aiding future maintainers.

### 10. **Enhanced Error Messages**
Make error handling more informative for troubleshooting:
```python
def download_pdf(self, pdf_url: str, save_path: str):
    try:
        urllib.request.urlretrieve(pdf_url, save_path)
    except urllib.error.URLError as e:
        logger.error(f"Network error occurred while downloading {pdf_url}: {e}")
        raise
    except OSError as e:
        logger.error(f"File save error for {save_path}: {e}")
        raise
```

## Security Considerations
1. **URL Validation**: Ensure that all URLs being fetched are validated to prevent malicious downloads.
2. **Timeouts**: Implement timeouts in network requests to improve robustness against connectivity issues.
3. **File Path Validation**: Mitigate risks of directory traversal attacks by validating paths provided for saving files.

## Performance Considerations
1. **Asynchronous Handling**: Consider adopting async methodologies for trying multiple downloads concurrently to improve performance for heavy workloads.
2. **Caching Mechanisms**: Implement caching for repeated requests to enhance speed and reduce load on the ArXiv server.
3. **Retry Logic**: Introduce retry mechanisms for failed requests for better fault tolerance.

## Conclusion
The `arxiv_paper_tool.py` serves as a valuable tool for users accessing academic papers. Implementing the recommendations outlined above will substantially enhance functionality, security, and maintainability. Future modifications should also include documentation and consistent code review practices to uphold quality.

Please review these suggestions and consider implementing the proposed changes to ensure a robust and reliable tool for the community.

HarikrishnanK9 commented:
@joaomdmoura Updated as per your review

HarikrishnanK9 commented:
Hi @joaomdmoura  Just a quick reminder in case you forgot.I’ve updated  PR based on the feedback.Let me know when you get a chance to review it. @lucasgomide tagging you as well in case you're available to help review. Thanks Crewai Team.

HarikrishnanK9 commented:
Hi @joaomdmoura, Just a quick reminder in case you forgot. I’ve updated the PR based on the feedback. Let me know when you get a chance to review it. @lucasgomide tagging you as well in case you're available to help review. Thanks, Crewai Team.

HarikrishnanK9 commented:
@lucasgomide @joaomdmoura  I am waiting for a review and approval . Please let me know if any changes or updates are required from my side.

lucasgomide commented on crewai_tools/tools/arxiv_paper_tool/arxiv_paper_tool.py:
```diff
@@ -0,0 +1,137 @@
+from typing import ClassVar
+import logging
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
```
drop the `logging.basicConfig(level=logging.INFO)`. You can keep the `logger = logging..` 

```suggestion
logger = logging.getLogger(__name__)
```

We recently discovered a bug where excessive logs appeared during Crew execution because of this setup. Setting logging.basicConfig(level=...) overrides the logging configuration for all loggers, including those from your Crews, flooding your execution with unwanted logs.

HarikrishnanK9 replied:
updated

lucasgomide commented on crewai_tools/tools/arxiv_paper_tool/arxiv_paper_tool.py:
```diff
@@ -0,0 +1,137 @@
+    def fetch_arxiv_data(self, search_query: str, max_results: int) -> List[dict]:
+        api_url = f"{self.BASE_API_URL}?search_query={urllib.parse.quote(search_query)}&start=0&max_results={max_results}"
+        logger.info(f"Fetching data from Arxiv API: {api_url}")
+        data = urllib.request.urlopen(api_url).read().decode('utf-8')
```
use context manager to ensure proper cleanup; also specify a atimeout value

```python
with urllib.request.urlopen(request, timeout=REQUEST_TIMEOUT) as response:
```

lucasgomide commented on crewai_tools/tools/arxiv_paper_tool/arxiv_paper_tool.py:
```diff
@@ -0,0 +1,137 @@
+    def fetch_arxiv_data(self, search_query: str, max_results: int) -> List[dict]:
+        api_url = f"{self.BASE_API_URL}?search_query={urllib.parse.quote(search_query)}&start=0&max_results={max_results}"
+        logger.info(f"Fetching data from Arxiv API: {api_url}")
+        data = urllib.request.urlopen(api_url).read().decode('utf-8')
```
Another question: should we ensure the respose is OK before handler?
Something like this should be enough:
```python
                    if response.status != 200:
                        raise Exception(f"HTTP {response.status}: {response.reason}")
```

HarikrishnanK9 replied:
resolved

HarikrishnanK9 commented:
@lucasgomide Apologies for the delayed response. Thank you for your valuable time and for the detailed review and instructions. I will complete the changes as soon as possible. Once everything is finalised on my end, I will inform you for the final review.

HarikrishnanK9 commented on crewai_tools/tools/arxiv_paper_tool/arxiv_paper_tool.py:
```diff
@@ -0,0 +1,137 @@
+    max_results: int = Field(5, ge=1, le=100, description="Max results to fetch; must be between 1 and 100")
+    download_pdfs: Optional[bool] = Field(False, description="If True, download PDFs to local folder")
+    save_dir: Optional[str] = Field("./arxiv_pdfs", description="Directory path to save downloaded PDFs")
+    use_title_as_filename: Optional[bool] = Field(False, description="Use paper title as PDF filename instead of arXiv ID")
```
updated

HarikrishnanK9 commented on crewai_tools/tools/arxiv_paper_tool/arxiv_paper_tool.py:
```diff
@@ -0,0 +1,137 @@
+    description: str = "Fetches metadata from Arxiv based on a search query and optionally downloads PDFs."
+    args_schema: Type[BaseModel] = ArxivToolInput
+
+    def _run(self, **kwargs) -> str:
```
updated

lucasgomide commented:
@HarikrishnanK9 hey! is it ready for review?

HarikrishnanK9 commented:
@lucasgomide I've addressed the feedback and made the necessary updates. I didn’t inform you earlier as I hadn’t included any examples using the tool. Could you please review the updated version? If it’s not fully aligned with CrewAI requirements, I’m happy to revise it accordingly.  Apologies for the delay — I was unavailable due to some personal commitments. I'm currently available and also working in parallel to add a few more tools and examples

HarikrishnanK9 commented:
Hi @lucasgomide, just following up on this PR. Let me know if any changes are needed .

HarikrishnanK9 commented:
@lucasgomide syncing the branch resolved those issues.

HarikrishnanK9 commented:
Hi @lucasgomide, I just wanted to follow up one more time on this PR. All feedback has been addressed. Thanks again for your time
